### ISSUE TITLE
Problem while retraining.

### ISSUE THREAD

**Issue by sagar1899** [azavea/raster-vision#1216]
i'm running object detection file given by RV.

model 1 : my GeoJson file contains two objects car and tree. and in ClassConfig method contains
                ClassConfig = {'car', 'tree', 'building'}. i trained model in "model1.pth" file.

then

i give GeoJson file contains one object "building". , and ClassConfig = {'car', 'tree', 'building'} is like this.
while training this model i gave init_weights="model1.pth" (previous pretrained model) and trained "model2.pth" model.

while predicting data with "model2.pth" file , only 'building' has been predicted. car and tree is not predicted.

what is the solution ?

**Comment by AdeelH** [comment]
This is probably a case of catastrophic forgetting (https://en.wikipedia.org/wiki/Catastrophic_interference). When you re-train the model on buildings only, it forgets how to detect cars and trees and only focuses on buildings. 

It would be better to train on all classes at the same time. If both of your GeoJSON files describe objects in the same image, you should combine them into one GeoJSON file.

**Comment by sagar1899** [comment]
> This is probably a case of catastrophic forgetting (https://en.wikipedia.org/wiki/Catastrophic_interference). When you re-train the model on buildings only, it forgets how to detect cars and trees and only focuses on buildings only.
> 
> It would be better to train on all classes at the same time. If both of your GeoJSON files describe objects in the same image, you should combine them into one GeoJSON file.

i'm facing same problem in retraining with same class.

i have trained one model with ClassConfig={vehicle} that contains data of "car" and save in "model.pth"
and while retraining i give ClassConfig={vehicle} and GeoJson label file contains info about "plane" and give init_weights="model.pth" and trained model model2.pth.

but model2.pth predict data of only "plane". it is not predict data of car.

what is the solution of this problem ?

**Comment by AdeelH** [comment]
Your problem is not clear to me. Please provide 
- a clearer and more complete description of what you are trying to achieve
- the entire `get_config()` for each run
- a sample of the geojson used in each run
- the commands that you used for executing each run
- the complete log of each run
- a description of what you expected the output to be
- a description of what the output actually was

**Comment by sagar1899** [comment]
1. My Get_Config function for first training.

	1- input image 1.tif and 2.tif contains data of "car" and 1.geojson and 2.geojson files contains polygons where car is locate.
	
	2- i have trained this data and save model in "model1.pth" file.
	
	3- my class config function is "class_config = ClassConfig(names=['vehicle'])"

	def get_config(runner, raw_uri, processed_uri, root_uri, test=False):
    train_ids = [
        '1', '2'
    ]
    val_ids = ['11']

    if test:
        train_ids = train_ids[0:1]
        val_ids = val_ids[0:1]

    def make_scene(id):
        raster_uri = join(raw_uri,
                          {}.tif'.format(id))
        label_uri = join(processed_uri,{}.geojson'.format(id))

        if test:
            crop_uri = join(processed_uri, 'crops',
                            os.path.basename(raster_uri))
            save_image_crop(
                raster_uri,
                crop_uri,
                label_uri=label_uri,
                size=1000,
                min_features=5)
            raster_uri = crop_uri

        raster_source = RasterioSourceConfig(
            uris=[raster_uri], channel_order=[0, 1, 2])

        label_source = ObjectDetectionLabelSourceConfig(
            vector_source=GeoJSONVectorSourceConfig(
                uri=label_uri, default_class_id=0, ignore_crs_field=True))

        return SceneConfig(
            id=id, raster_source=raster_source, label_source=label_source)

    class_config = ClassConfig(names=['vehicle'])
    chip_sz = 300
    dataset = DatasetConfig(
        class_config=class_config,
        train_scenes=[make_scene(id) for id in train_ids],
        validation_scenes=[make_scene(id) for id in val_ids])
    chip_options = ObjectDetectionChipOptions(neg_ratio=5.0, ioa_thresh=0.9)
    predict_options = ObjectDetectionPredictOptions(
        merge_thresh=0.5, score_thresh=0.9)

    backend = PyTorchObjectDetectionConfig(
        model=ObjectDetectionModelConfig(backbone=Backbone.resnet18),
        solver=SolverConfig(
            lr=1e-4,
            num_epochs=10,
            test_num_epochs=2,
            batch_sz=16,
            one_cycle=True),
        log_tensorboard=True,
        run_tensorboard=False,
        test_mode=test)

    return ObjectDetectionConfig(
        root_uri=root_uri,
        dataset=dataset,
        backend=backend,
        train_chip_sz=chip_sz,
        predict_chip_sz=chip_sz,
        chip_options=chip_options,
        predict_options=predict_options)
		
2. then i want to train model1.pth with different data solver

	1- input image 3.tif and 4.tif contains data of "plane" and 3.geojson and 4.geojson files contains polygons where plane is locate.
	
	2- i have trained this data and save model in "model2.pth" file.
	
	3- my class config function is "class_config = ClassConfig(names=['vehicle'])"

	def get_config(runner, raw_uri, processed_uri, root_uri, test=False):
    train_ids = [
        '3', '4'
    ]
    val_ids = ['11']

    if test:
        train_ids = train_ids[0:1]
        val_ids = val_ids[0:1]

    def make_scene(id):
        raster_uri = join(raw_uri,
                          {}.tif'.format(id))
        label_uri = join(processed_uri,{}.geojson'.format(id))

        if test:
            crop_uri = join(processed_uri, 'crops',
                            os.path.basename(raster_uri))
            save_image_crop(
                raster_uri,
                crop_uri,
                label_uri=label_uri,
                size=1000,
                min_features=5)

            raster_uri = crop_uri

        raster_source = RasterioSourceConfig(
            uris=[raster_uri], channel_order=[0, 1, 2])

        label_source = ObjectDetectionLabelSourceConfig(
            vector_source=GeoJSONVectorSourceConfig(
                uri=label_uri, default_class_id=0, ignore_crs_field=True))

        return SceneConfig(
            id=id, raster_source=raster_source, label_source=label_source)

    class_config = ClassConfig(names=['vehicle'])
    chip_sz = 300
    dataset = DatasetConfig(
        class_config=class_config,
        train_scenes=[make_scene(id) for id in train_ids],
        validation_scenes=[make_scene(id) for id in val_ids])
    chip_options = ObjectDetectionChipOptions(neg_ratio=5.0, ioa_thresh=0.9)
    predict_options = ObjectDetectionPredictOptions(
        merge_thresh=0.5, score_thresh=0.9)

    backend = PyTorchObjectDetectionConfig(
        model=ObjectDetectionModelConfig(backbone=Backbone.resnet18, init_weights="path to model1.pth"),
        solver=SolverConfig(
            lr=1e-4,
            num_epochs=10,
            test_num_epochs=2,
            batch_sz=16,
            one_cycle=True),
        log_tensorboard=True,
        run_tensorboard=False,
        test_mode=test)

    return ObjectDetectionConfig(
        root_uri=root_uri,
        dataset=dataset,
        backend=backend,
        train_chip_sz=chip_sz,
        predict_chip_sz=chip_sz,
        chip_options=chip_options,
        predict_options=predict_options)
		
		
3- then i predict another image with model2.pth model file.
	it is predict only plane type data
	it does not predict car data.
	
4- **i want that model2.pth will predict both "car" as well "plane".** 

5- command is used is 
       "python -m rastervision.pipeline.cli run_command "path to pipeline-config.json" chip"

        "python -m rastervision.pipeline.cli run_command "path to pipeline-config.json" train"

       "python -m rastervision.pipeline.cli run_command "path to pipeline-config.json" predict"
   
for both training.


**i want to retrain my model and predict both types of data taken while training.**

**Comment by AdeelH** [comment]
- Is there a reason you want to train the vehicle and plane classes separately? You can set the `class_config` to `ClassConfig(names=['vehicle', 'plane'])` and pass in all 4 images in a single training run.
- I am not sure how your 2nd model is predicting planes. You did not specify `'plane'` in the class config, so it should not even have been able to extract plane labels from the geojson files.
- The correct way to go about what you are trying to do would be to use `ClassConfig(names=['vehicle', 'plane'])` in both training runs. This will not necessarily produce a model that is good at detecting both classes (see my first comment in this issue).

P.S. Please put in the time and effort to format your posts/comments properly. You can learn more about GitHub markdown here: https://guides.github.com/features/mastering-markdown.

**Comment by sagar1899** [comment]
> * Is there a reason you want to train the vehicle and plane classes separately? You can set the `class_config` to `ClassConfig(names=['vehicle', 'plane'])` and pass in all 4 images in a single training run.
> * I am not sure how your 2nd model is predicting planes. You did not specify `'plane'` in the class config, so it should not even have been able to extract plane labels from the geojson files.
> * The correct way to go about what you are trying to do would be to use `ClassConfig(names=['vehicle', 'plane'])` in both training runs. This will not necessarily produce a model that is good at detecting both classes (see my first comment in this issue).
> 
> P.S. Please put in the time and effort to format your posts/comments properly. You can learn more about GitHub markdown here: https://guides.github.com/features/mastering-markdown.
I understand your concern but my problem is while retraining a pre-trained model my new model is not contained data of pre-trained model.

simply,  i have model "model1.pth" that predict car. while retraining that model with new data of plane and saves model in "model2.pth".
but model2.pth is only predicts plane. it is not predict car.

i remember that in one comments you mention that there is problem called catastrophic forgetting.
what is the solution to that?

my concern is that i want to train pre-trained model and that new model can predict both old data as well as new data.

I want to continusely train a one model with new datasets.

**Comment by AdeelH** [comment]
>simply, i have model "model1.pth" that predict car. while retraining that model with new data of plane and saves model in >"model2.pth".
>but model2.pth is only predicts plane. it is not predict car.

I have described the correct way to do this in my last comment.

>i remember that in one comments you mention that there is problem called catastrophic forgetting.
what is the solution to that?
>
>my concern is that i want to train pre-trained model and that new model can predict both old data as well as new data.
>
>I want to continusely train a one model with new datasets.

This a general machine learning question (i.e. not specific to RV) that might be better suited for some other forums.
